render/render_test.go

```diff
@@ -18,73 +18,161 @@ const (
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port:-9090}:80"
+`)
+	parameters := strings.NewReader(`
+front:
+  port: 8080
+`)
+	app := &types.App{Path: "my-app"}
+	assert.NilError(t, types.Metadata(metadata)(app))
+	assert.NilError(t, types.WithComposes(composeFile)(app))
+	assert.NilError(t, types.WithParameters(parameters)(app))
+	userParameters := map[string]string{
+		"front.deploy.replicas": "9",
+		"front.port":            "4242",
+		"back.port":             "6666",
 	}
-	_, err := render("foo.dockerapp", configFiles, finalEnv, nil)
-	assert.Check(t, err != nil)
-	assert.Check(t, is.ErrorContains(err, "required variable"))
+	_, err := Render(app, userParameters, nil)
+	assert.ErrorContains(t, err, "Parameters must not have default values set in compose file.")
 }
 
-func TestRender(t *testing.T) {
-	configFiles := []composetypes.ConfigFile{
-		{
-			Config: map[string]interface{}{
-				"version": "3",
-				"services": map[string]interface{}{
-					"foo": map[string]interface{}{
-						"image":   "busybox:${version}",
-						"command": []interface{}{"-text", "${foo.bar}"},
-					},
-				},
-			},
-		},
+func TestSubstituteMixedParams(t *testing.T) {
+	composeFile := `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "${front.port}:80"
+    deploy:
+      replicas: ${front-deploy.replicas}
+  back:
+    ports:
+     - "$back.port:90"
+`
+	parameters := map[string]string{
+		"front.port":            "8080",
+		"back.port":             "9000",
+		"front-deploy.replicas": "3",
 	}
-	finalEnv := map[string]string{
-		"version": "latest",
-		"foo.bar": "baz",
+	s, err := substituteParams(parameters, composeFile)
+	assert.NilError(t, err)
+	assert.Equal(t, s, `
+version: "3.6"
+services:
+  front:
+    ports:
+     - "8080:80"
+    deploy:
+      replicas: 3
+  back:
+    ports:
+     - "9000:90"
+`)
+}
+
+func TestSubstituteSkipDoubleDollarCase(t *testing.T) {
+	composeFile := `
+	version: "3.7"
+	services:
+	  front:
+		command: monitor --source ${app.name}-$app.version $$dollar
+`
+	parameters := map[string]string{
+		"app.name":    "test",
+		"app.version": "version",
 	}
-	c, err := render("foo.dockerapp", configFiles, finalEnv, nil)
+	s, err := substituteParams(parameters, composeFile)
 	assert.NilError(t, err)
-	assert.Check(t, is.Len(c.Services, 1))
-	assert.Check(t, is.Equal(c.Services[0].Image, "busybox:latest"))
-	assert.Check(t, is.DeepEqual([]string(c.Services[0].Command), []string{"-text", "baz"}))
+	assert.Equal(t, s, `
+	version: "3.7"
+	services:
+	  front:
+		command: monitor --source test-version $$dollar
+`)
+}
+
+func TestSubstituteMissingParameterValue(t *testing.T) {
+	composeFile := `
+	version: "3.7"
+	services:
+	  front:
+		deploy:
+		  replicas: ${myapp.nginx_replicas}
+	  debug:
+		ports:
+		- $aport
+`
+	parameters := map[string]string{
+		"aport": "10000",
+	}
+	_, err := substituteParams(parameters, composeFile)
+	assert.ErrorContains(t, err, "Value not found in parameters.")
```

Comment:
Can you assert the whole error message, with the right parameter name? Just to be sure it's the good one which is missing 😅